File: various/src/main/java/com/chrnie/various/ItemMatcher.java
package com.chrnie.various;

import android.support.annotation.NonNull;
import java.util.List;

public interface ItemMatcher {

  int getViewType(@NonNull Object date);

  @NonNull
  Item getItem(int viewType);

  interface Factory {
    @NonNull
    ItemMatcher create(@NonNull List<Item> itemList);
  }
}

File: various/src/main/java/com/chrnie/various/ItemMatcher.kt
package com.chrnie.various

import androidx.recyclerview.widget.RecyclerView

interface ItemMatcher {

    fun requestViewType(date: Any): Int

    fun requestViewHolderBinder(viewType: Int): ViewBinder<Any, RecyclerView.ViewHolder>

    interface Factory {
        fun create(itemList: List<Item<*, *>>): ItemMatcher
    }
}